pkg/graph/builder.go

```diff
@@ -672,13 +684,125 @@ func extractDependencies(env *cel.Env, expression string, resourceNames []string
 	return dependencies, isStatic, nil
 }
 
+// setExpectedTypeOnDescriptor sets the ExpectedType field on a FieldDescriptor.
+// This is the single place where ExpectedType is determined for all field descriptors.
+//
+// For string templates (multiple expressions like "foo-${expr1}-${expr2}"):
+//   - Always sets to cel.StringType since concatenation produces strings
+//
+// For standalone expressions (single expression like "${expr}"):
+//  1. Parses path into segments (e.g., "spec.containers[0].name" -> ["spec", "containers", [0], "name"])
+//  2. Walks through each segment, building type name and navigating schema:
+//     - Named segments: append to type name, look up in schema
+//     - Index segments: dereference array to element schema, append ".@idx" to type name
+//  3. Converts final schema to CEL type with constructed type name
+func resolveSchemaAndTypeName(segments []fieldpath.Segment, rootSchema *spec.Schema, resourceID string) (*spec.Schema, string, error) {
+	typeName := krocel.TypeNamePrefix + resourceID
+	currentSchema := rootSchema
+
+	for _, seg := range segments {
+		if seg.Name != "" {
+			typeName = typeName + "." + seg.Name
+			currentSchema = lookupSchemaAtPath(currentSchema, seg.Name)
+			if currentSchema == nil {
+				return nil, "", fmt.Errorf("field %q not found in schema", seg.Name)
+			}
+		}
+
+		if seg.Index != -1 {
+			if currentSchema.Items != nil && currentSchema.Items.Schema != nil {
+				currentSchema = currentSchema.Items.Schema
+				typeName = typeName + ".@idx"
+			} else {
+				return nil, "", fmt.Errorf("field is not an array")
+			}
+		}
+	}
+
+	return currentSchema, typeName, nil
+}
+
+func setExpectedTypeOnDescriptor(descriptor *variable.FieldDescriptor, rootSchema *spec.Schema, resourceID string) {
+	if !descriptor.StandaloneExpression {
+		descriptor.ExpectedType = cel.StringType
+		return
+	}
+
+	segments, err := fieldpath.Parse(descriptor.Path)
+	if err != nil {
+		descriptor.ExpectedType = cel.DynType
+		return
+	}
+
+	schema, typeName, err := resolveSchemaAndTypeName(segments, rootSchema, resourceID)
+	if err != nil {
+		descriptor.ExpectedType = cel.DynType
+		return
+	}
+
+	descriptor.ExpectedType = getCelTypeFromSchema(schema, typeName)
+}
+
+// getCelTypeFromSchema converts an OpenAPI schema to a CEL type with the given type name
+func getCelTypeFromSchema(schema *spec.Schema, typeName string) *cel.Type {
+	if schema == nil {
+		return cel.DynType
+	}
+
+	declType := openapi.SchemaDeclType(schema, false)
```

Comment:
I think this is an inconsistent schema decl type since we now have our own

Reply:
Good catch!